typing/typemod.ml

```diff
@@ -150,14 +153,144 @@ let make p n i =
   let open Variance in
   set May_pos p (set May_neg n (set May_weak n (set Inj i null)))
 
+let rec iter_path_apply p ~f =
+  match p with
+  | Pident _ -> ()
+  | Pdot (p, _, _) -> iter_path_apply p ~f
+  | Papply (p1, p2) ->
+     iter_path_apply p1 ~f;
+     iter_path_apply p2 ~f;
+     f p1 p2 (* after recursing, so we know both paths are well typed *)
+
+let path_is_prefix =
+  let rec list_is_prefix ~strict l ~prefix =
+    match l, prefix with
+    | [], [] -> not strict
+    | _ :: _, [] -> true
+    | [], _ :: _ -> false
+    | s1 :: t1, s2 :: t2 ->
+       String.equal s1 s2 && list_is_prefix ~strict t1 ~prefix:t2
+  in
+  fun ~strict path ~prefix ->
+    match Path.flatten path, Path.flatten prefix with
+    | `Contains_apply, _ | _, `Contains_apply -> false
+    | `Ok (ident1, l1), `Ok (ident2, l2) ->
+       Ident.same ident1 ident2
+       && list_is_prefix ~strict l1 ~prefix:l2
+
+let iterator_with_env env =
+  let env = ref env in
+  let super = Btype.type_iterators in
+  env, { super with
+    Btype.it_signature = (fun self sg ->
+      let env_before = !env in
+      List.iter (fun i -> env := Env.add_item i !env) sg;
+      super.Btype.it_signature self sg;
+      env := env_before
+    );
+    Btype.it_module_type = (fun self -> function
+    | Mty_functor (param, mty_arg, mty_body) ->
+      may (self.Btype.it_module_type self) mty_arg;
+      let env_before = !env in
+      env := Env.add_module ~arg:true param (Btype.default_mty mty_arg) !env;
+      self.Btype.it_module_type self mty_body;
+      env := env_before;
+    | mty ->
+      super.Btype.it_module_type self mty
+    )
+  }
+
+let retype_applicative_functor_type ~loc env funct arg =
+  let mty_functor = (Env.find_module funct env).md_type in
+  let mty_arg = (Env.find_module arg env).md_type in
+  let mty_param =
+    match Env.scrape_alias env mty_functor with
+    | Mty_functor (_, Some mty_param, _) -> mty_param
+    | _ -> assert false
```

Comment:
Depressingly this `assert false` can be triggered due to MPR#7611. I wouldn't bother changing it, but thought I should probably mention it.